kclvm/compiler_base/3rdparty/rustc_errors/src/styled_buffer.rs

```diff
@@ -0,0 +1,146 @@
+    lines: Vec<Vec<StyledChar>>,
+}
+
+#[derive(Clone)]
+struct StyledChar {
+    chr: char,
+    style: Option<Box<dyn Style>>,
+}
+
+pub struct StyledString {
+    pub text: String,
+    pub style: Option<Box<dyn Style>>,
+}
+
+impl StyledChar {
+    const SPACE: StyledChar = StyledChar::new(' ', None);
+
+    const fn new(chr: char, style: Option<Box<dyn Style>>) -> Self {
+        StyledChar { chr, style }
+    }
+}
+
+impl StyledBuffer {
+    pub fn new() -> StyledBuffer {
+        StyledBuffer { lines: vec![] }
+    }
+
+    /// Returns content of `StyledBuffer` split by lines and line styles
+    pub fn render(&self) -> Vec<Vec<StyledString>> {
+        let mut output: Vec<Vec<StyledString>> = vec![];
+        let mut styled_vec: Vec<StyledString> = vec![];
+
+        for styled_line in &self.lines {
+            let mut current_style = None;
+            let mut current_text = String::new();
+
+            for sc in styled_line {
+                if sc.style != current_style {
+                    if !current_text.is_empty() {
+                        styled_vec.push(StyledString {
+                            text: current_text,
+                            style: current_style,
+                        });
+                    }
+                    current_style = sc.style.clone();
+                    current_text = String::new();
+                }
+                current_text.push(sc.chr);
+            }
+            if !current_text.is_empty() {
+                styled_vec.push(StyledString {
+                    text: current_text,
+                    style: current_style,
+                });
+            }
+
+            // done with the row, push and keep going
+            output.push(styled_vec);
+
+            styled_vec = vec![];
+        }
+
+        output
+    }
+
+    fn ensure_lines(&mut self, line: usize) {
+        if line >= self.lines.len() {
+            self.lines.resize(line + 1, Vec::new());
+        }
+    }
+
+    /// Sets `chr` with `style` for given `line`, `col`.
+    /// If `line` does not exist in our buffer, adds empty lines up to the given
+    /// and fills the last line with unstyled whitespace.
+    pub fn putc(&mut self, line: usize, col: usize, chr: char, style: Option<Box<dyn Style>>) {
+        self.ensure_lines(line);
+        if col >= self.lines[line].len() {
+            self.lines[line].resize(col + 1, StyledChar::SPACE);
+        }
+        self.lines[line][col] = StyledChar::new(chr, style);
+    }
+
+    /// Sets `string` with `style` for given `line`, starting from `col`.
+    /// If `line` does not exist in our buffer, adds empty lines up to the given
+    /// and fills the last line with unstyled whitespace.
+    pub fn puts(&mut self, line: usize, col: usize, string: &str, style: Option<Box<dyn Style>>) {
+        let mut n = col;
+        for c in string.chars() {
+            self.putc(line, n, c, style.clone());
+            n += 1;
+        }
+    }
+
```

Comment:
Please add docstring for newly-added putl() and appendl(), just like any other functions.

Reply:
rename `putl()` to `pushs()`.
rename `appendl()` to `append_last_line()`.
add some docstring.